Fold IOSurfaceImageBuffer into ImageBuffer

hortont424 opened the pull request:
#### b9368743dc7a35dcd9f8d53113162bcb18d0f6f3
<pre>
Fold IOSurfaceImageBuffer into ImageBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=265698">https://bugs.webkit.org/show_bug.cgi?id=265698</a>
<a href="https://rdar.apple.com/119050705">rdar://119050705</a>

Reviewed by Said Abou-Hallawa.

IOSurfaceImageBuffer complicates adding new non-backend-specific ImageBuffer
subclasses, which seems more valuable than using a whole subclass to expose
backend-specific methods.

Remove it in favor of a failable getter on ImageBuffer.

Also fix up a questionable cast (assuming all accelerated image buffers
contain ImageBufferIOSurfaceBackend if IOSurface is enabled seems ill-advised).

* Source/WTF/wtf/cocoa/TollFreeBridging.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::surface):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::isImageBufferIOSurfaceBackend const):
* Source/WebCore/platform/graphics/cg/IOSurfaceImageBuffer.h: Removed.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
(isType):
* Source/WebCore/platform/graphics/coreimage/FilterImageCoreImage.mm:
(WebCore::FilterImage::imageBufferFromCIImage):
* Source/WebCore/platform/graphics/coreimage/SourceGraphicCoreImageApplier.mm:
(WebCore::SourceGraphicCoreImageApplier::apply const):

Canonical link: <a href="https://commits.webkit.org/271509@main">https://commits.webkit.org/271509@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc2c67114b534aed57d51f2a2c9db67b38239130

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28619 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29999 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9379 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4632 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/31256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6004 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24610 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31256 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25611 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/24757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26204 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26057 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/32587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/27701 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5335 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3507 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6944 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35237 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6849 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5799 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7611 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WTF/wtf/cocoa/TollFreeBridging.h:
```diff
@@ -57,7 +57,7 @@ WTF_DECLARE_TOLL_FREE_BRIDGING_TRAITS(CFSet, NSSet)
 WTF_DECLARE_TOLL_FREE_BRIDGING_TRAITS(CFTimeZone, NSTimeZone)
 WTF_DECLARE_TOLL_FREE_BRIDGING_TRAITS(CFURL, NSURL)
-WTF_DECLARE_TOLL_FREE_BRIDGING_TRAITS(IOSurface, IOSurface)
+WTF_DECLARE_TOLL_FREE_BRIDGING_TRAITS(IOSurface, ::IOSurface)
```
What does this change do?

hortont424 replied:
disambiguates IOSurface the ObjC class from WebCore::IOSurface (and fixes the build :D)

shallawa commented on Source/WebCore/platform/graphics/ImageBuffer.cpp:
```diff
@@ -358,6 +358,15 @@ RefPtr<NativeImage> ImageBuffer::filteredNativeImage(Filter& filter, Function<vo
+        return backend->surface();
+    return nullptr;
+}
+#endif
```
If we follow the pattern of `createCairoSurface()` we will not need the casting. The `ImageBufferBackend` will have a virtual method called `surface()` also.
```
IOSurface* ImageBuffer::surface()
{
    return m_backend ? m_backend-> surface() : nullptr;
}
```

hortont424 replied:
Could do that too. Which do we prefer!

shallawa replied:
I do not know but I just care about consistency here.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -43,6 +43,11 @@ namespace WebCore {
+bool ImageBufferIOSurfaceBackend::isImageBufferIOSurfaceBackend() const
+{
+    return true;
+}
```
I think we usually put the virtual isXYZ() in the header file.

hortont424 replied:
Yes, but the stylebot goes crazy about inline functions in WEBCORE_EXPORT classes. Maybe you just ignore it??

shallawa replied:
Oh yes I forgot the WEBCORE_EXPORT on the class level.

shallawa commented on Source/WebCore/platform/graphics/coreimage/FilterImageCoreImage.mm:
```diff
@@ -61,12 +61,12 @@
 
     auto destRect = FloatRect { FloatPoint(), m_absoluteImageRect.size() };
-    [sharedCIContext().get() render: m_ciImage.get() toIOSurface: downcast<IOSurfaceImageBuffer>(*m_imageBuffer).surface().surface() bounds:destRect colorSpace:m_colorSpace.platformColorSpace()];
+    [sharedCIContext().get() render: m_ciImage.get() toIOSurface: m_imageBuffer->surface()->surface() bounds:destRect colorSpace:m_colorSpace.platformColorSpace()];
```
If `ImageBuffer::create()` returns a valid `ImageBuffer`, it is guaranteed the IOSurface was created. But maybe it will be clearer if we add an assertion saying this before this statement.

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
```diff
@@ -50,6 +50,8 @@ class WEBCORE_EXPORT ImageBufferIOSurfaceBackend : public ImageBufferCGBackend {
     
     static constexpr RenderingMode renderingMode = RenderingMode::Accelerated;
 
+    bool isImageBufferIOSurfaceBackend() const override;
```
We have four backends which are backed by an IOSurface 

1. ImageBufferIOSurfaceBackend
2. ImageBufferShareableMappedIOSurfaceBackend
3. ImageBufferRemoteIOSurfaceBackend
4. ImageBufferShareableMappedIOSurfaceBitmapBackend

Only the first two will return true for this function. Is this what we want? I am just confused by ImageBufferShareableMappedIOSurfaceBackend and I do not whether it should return true or not.

hortont424 replied:
... good question.

hortont424 replied:
I think it would be definitely wrong to say yes from _this_ (the is-castable function).

I also think it's not necessary to make surface() work for the later two, simply because the old mechanism wouldn't have covered them either?

hortont424 committed:
Fold IOSurfaceImageBuffer into ImageBuffer
https://bugs.webkit.org/show_bug.cgi?id=265698
rdar://119050705

Reviewed by Said Abou-Hallawa.

IOSurfaceImageBuffer complicates adding new non-backend-specific ImageBuffer
subclasses, which seems more valuable than using a whole subclass to expose
backend-specific methods.

Remove it in favor of a failable getter on ImageBuffer.

Also fix up a questionable cast (assuming all accelerated image buffers
contain ImageBufferIOSurfaceBackend if IOSurface is enabled seems ill-advised).

* Source/WTF/wtf/cocoa/TollFreeBridging.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::create):
(WebCore::ImageBuffer::surface):
* Source/WebCore/platform/graphics/ImageBuffer.h:
* Source/WebCore/platform/graphics/ImageBufferBackend.h:
(WebCore::ImageBufferBackend::isImageBufferIOSurfaceBackend const):
* Source/WebCore/platform/graphics/cg/IOSurfaceImageBuffer.h: Removed.
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
(isType):
* Source/WebCore/platform/graphics/coreimage/FilterImageCoreImage.mm:
(WebCore::FilterImage::imageBufferFromCIImage):
* Source/WebCore/platform/graphics/coreimage/SourceGraphicCoreImageApplier.mm:
(WebCore::SourceGraphicCoreImageApplier::apply const):

Canonical link: https://commits.webkit.org/271509@main
